File: Construct_TREE_inandlevel/construct_tree.py
class TreeNode:
    def __init__(self, value):
        self.val = value
        self.left = None
        self.right = None

def build_tree_helper(inorder, levelorder, start, end, index_map):
    if start > end:
        return None

    root = TreeNode(levelorder[0])
    root_index = index_map[root.val]

    left_levelorder = [node for node in levelorder[1:] if index_map[node] < root_index]
    right_levelorder = [node for node in levelorder[1:] if index_map[node] > root_index]

    root.left = build_tree_helper(inorder, left_levelorder, start, root_index - 1, index_map)
    root.right = build_tree_helper(inorder, right_levelorder, root_index + 1, end, index_map)

    return root

def build_tree(inorder, levelorder):
    index_map = {val: i for i, val in enumerate(inorder)}
    return build_tree_helper(inorder, levelorder, 0, len(inorder) - 1, index_map)

def print_inorder(root):
    if root:
        print_inorder(root.left)
        print(root.val, end=" ")
        print_inorder(root.right)

if __name__ == "__main__":
    inorder = [4, 8, 10, 12, 14, 20, 22]
    levelorder = [20, 8, 22, 4, 12, 10, 14]

    root = build_tree(inorder, levelorder)

    print("Inorder traversal of the constructed tree:")
    print_inorder(root)

File: Construct_TREE_inandlevel/construct_tree.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int value) {
        val = value;
        left = nullptr;
        right = nullptr;
    }
};

TreeNode* buildTreeHelper(vector<int>& inorder, vector<int>& levelorder, int start, int end, unordered_map<int, int>& map) {
    if (start > end)
        return nullptr;

    TreeNode* root = new TreeNode(levelorder[0]);
    int rootIndex = map[root->val];

    vector<int> leftLevelOrder, rightLevelOrder;
    for (int i = 1; i < levelorder.size(); ++i) {
        int index = map[levelorder[i]];
        if (index < rootIndex)
            leftLevelOrder.push_back(levelorder[i]);
        else if (index > rootIndex)
            rightLevelOrder.push_back(levelorder[i]);
    }

    root->left = buildTreeHelper(inorder, leftLevelOrder, start, rootIndex - 1, map);
    root->right = buildTreeHelper(inorder, rightLevelOrder, rootIndex + 1, end, map);

    return root;
}

TreeNode* buildTree(vector<int>& inorder, vector<int>& levelorder) {
    unordered_map<int, int> map;
    for (int i = 0; i < inorder.size(); ++i)
        map[inorder[i]] = i;

    return buildTreeHelper(inorder, levelorder, 0, inorder.size() - 1, map);
}

void printInorder(TreeNode* root) {
    if (root == nullptr)
        return;

    printInorder(root->left);
    cout << root->val << " ";
    printInorder(root->right);
}

int main() {
    vector<int> inorder = {4, 8, 10, 12, 14, 20, 22};
    vector<int> levelorder = {20, 8, 22, 4, 12, 10, 14};

    TreeNode* root = buildTree(inorder, levelorder);

    cout << "Inorder traversal of the constructed tree:\n";
    printInorder(root);
    cout << endl;

    return 0;
}
